Add option to display metadata

imendes93 opened the pull request:
This PR closes #27 .
It incorporates the requirements requested in #33, #45, #46, #42 and #44 , closing them.  

## Modification:

A new optional parameter has been added `--metadata`, that allows metadata in a csv format to be passed on to pcgr-nf **when** the `--csv` option is used. The help message has been update to reflect this change:

        nextflow run main.nf --help
        N E X T F L O W  ~  version 20.10.0
        Launching `main.nf` [small_jang] - revision: 8f1fcf9600
        Usage:
        The typical command for running the pipeline is as follows:
        nextflow run main.nf --vcf sample.vcf [Options]
        
        Essential paramenters:
          Single File Mode:
        --vcf           Input `VCF` file.
                      See example in `testdata/test.vcf`. 
        
          Multiple File Mode:
        --csv           A list of `VCF` files.
                      Should be a `*.csv` file with a header called `vcf` and the path to each file, one per line. 
                      See example in `testdata/list-vcf-files-local.csv`.
        
        PCGR Options:
        --pcgr_config   Tool config file (path)
                      (default: s3://fast-ngs/cloudos-public-data-ines/pcgr.toml)
        --pcgr_genome   Reference genome assembly
                      (default: grch38)
        --pcgr_data     URL for reference data bundle.
                      Optional filed. If not provided, the appropriate data bundle is infered from --pcgr_genome. 
                      (default: grch38)
        
        Optional paramenters:
        --metadata      `CSV` file with metadata information.
                      Available only when using --csv mode
                      Should be a `*.csv` file with the first column called `vcf` and the path to each file, one per line,
                      matching the file in --csv, followed by metadata, one per column, for each sample. 
                      The column `histological_type` is required to be present.
                      See example in `testdata/metadata.csv`.
        
        Resource Options:
        --max_cpus      Maximum number of CPUs (int)
                      (default: 16)  
        --max_memory    Maximum memory (memory unit)
                      (default: 20 GB)
        --max_time      Maximum time (time unit)
                      (default: 10d)
        See here for more info: https://github.com/lifebit-ai/pcgr-nf

The metadata is combined with the main `combined.tiers.tsv` output from PCGR, identified with the prefix `METADATA_` as to avoid possible conflict with existing columns and to facilitate search and query.  
![image](https://user-images.githubusercontent.com/73831087/130996606-87251902-191a-45f6-815a-5b21f8f0eecb.png)

As described in #40, the metadata is displayed in the gene summary table, containing basic information on each gene with variants
![image](https://user-images.githubusercontent.com/73831087/130994564-afb453c3-5f6b-4dce-812c-b0c72d33677e.png)
To this table it was also added the number of variants per tier
![image](https://user-images.githubusercontent.com/73831087/130994667-9628838b-5af2-4370-95ad-c85b2e3031bc.png)
In addition to the Number and Percentage of samples according to the categories defined in the `histological_type` column in the metadata. This, alognside `vcf`,  is a required column. 

|vcf     |sample_id|histological_type|sex   |library_name|
|--------|---------|-----------------|------|------------|
|test.vcf|G1       |Germinoma        |female|IGCT-1      |
|two.vcf |G2       |Mixed            |male  |IGCT-3      |

![image](https://user-images.githubusercontent.com/73831087/130995340-b77d2ce2-ba76-46c6-87ad-40a44b649b51.png)

## To test:

To test the pipeline with new test and reference data locally run:

    git clone https://github.com/lifebit-ai/pcgr-nf
    cd pcgr-nf
    git checkout metadata-dev
    NXF_VER=20.01.0 nextflow run . -profile standard,docker --config conf/test.config


### Runed locally 
![image](https://user-images.githubusercontent.com/73831087/130994114-74ac91ca-b41f-4e06-bdc4-71d31b8905ee.png)
![image](https://user-images.githubusercontent.com/73831087/130995593-e597a633-169a-4ea0-a300-d8962575badf.png)

### Runed on CloudOS
Public run with the latest commit id available at https://cloudos.lifebit.ai/public/jobs/6127bfc47db707019a0aab6a
![image](https://user-images.githubusercontent.com/73831087/130995802-4f02687a-e7af-48aa-961f-4d2423d62be6.png)
![image](https://user-images.githubusercontent.com/73831087/131003904-0bdb661f-ecbb-4c53-a64c-2017dc1f7f23.png)
![image](https://user-images.githubusercontent.com/73831087/131003963-7591bd66-8185-4531-8304-34d2a27bcb1f.png)

Vlad-Dembrovskyi reviewed (changes requested):
Tested locally - results looks correct and accurate, as per requirements from 13th Aug. Looks good!
Although I don't remember adding the Tier 1-4 cols to "By gene" table being one of the requirements, I really like it and think its a good idea. There is however a change we have to make before merging (see below).

![image](https://user-images.githubusercontent.com/64809705/131315501-2b1f4b66-34c4-47e9-b8c7-91d8085dc9b3.png)
![image](https://user-images.githubusercontent.com/64809705/131316062-206ff3e9-7974-4fb6-ae11-31d7fc7f5a17.png)
![image](https://user-images.githubusercontent.com/64809705/131316362-afe4e3aa-00e4-473c-b5f8-b11d0f735e23.png)

After reviewing the code changes I have one comment. I see that combine.py script now relies on argument position to retrieve metadata and other table files.
https://github.com/lifebit-ai/pcgr-nf/blob/21e268b59edf77fbfcbdbc51907f1744a6ef50bd/bin/combine.py#L10-L11
On the other hand I remember we were designing metadata to be optional (which I completely support) and metadata input channel is indeed configured to be optional here.
https://github.com/lifebit-ai/pcgr-nf/blob/21e268b59edf77fbfcbdbc51907f1744a6ef50bd/main.nf#L259
https://github.com/lifebit-ai/pcgr-nf/blob/21e268b59edf77fbfcbdbc51907f1744a6ef50bd/main.nf#L265-L266

This already leads to an internal conflict - if I want to run without metadata and theoretically don't provide metadata file at all - I will only provide table files as arguments to combine.py. I this case the first table will be always treated as metadata file, which is not correct and will certainly lead to errors when combining tables.
 
When I tried to run the pcgr-nf without metadata file and only one vcf provided with csv file I ran into an issue:
![image](https://user-images.githubusercontent.com/64809705/131319850-e0b21c39-41f5-4e91-a2c7-a06ea1bf2ba6.png)
![image](https://user-images.githubusercontent.com/64809705/131319889-b638547d-8b17-4664-a466-41f654985847.png)
![image](https://user-images.githubusercontent.com/64809705/131319912-5d93cfda-5729-43b1-a2c6-e44f8ad583f9.png)

Same error happened when I ran it without metadata but with two vcf files, as in the test profile.

I think there are two issues here:
1. Metadata file is always pasted by nextflow to the command even if channel is empty, as `input.2` in my case above. Possible solution: add line `metadata_param = params.metadata ? params.metadata : ""` under `script:` before script block starts `"""`, and then change `$metadata` to `$param_metadata` in the command line, so that metadata is only added when its not absent from input channel. This will save from having it as `input.2`.
2. Even with the solution above we will have problem that all arguments are positional and first one must always be metadata. So if we skip it - the first metadata table will be treated as metadata file and that may lead to other errors. I propose to make metadata an optional named parameter for combine.py script, and keep csv table arguments positional as they are now. In this case all positional args will be always expected as csv tables to combine, and metadata will be only expected to come from --metdata parameter if it is set. We will probably need to make more changes to combine.py script to make it truly optional with regarding to the metadata file.

Besides metadata file not being truly optional we also need to update the documentation in [docs/usage_cloudos.md](https://github.com/lifebit-ai/pcgr-nf/blob/metadata-dev/docs/usage_cloudos.md) to include new (at the moment required) parameter `--metadata`, and maybe additional explanations to output tables if necessary. Updated documentation must come with the same PR where new features are introduced.

imendes93 commented:
Thanks @Vlad-Dembrovskyi ! indeed there were some issues with the implementation of metadata as optional that slipped through my testing! I've addressed your requests:

- The metadata is now truly an optional argument, with pcgr-nf no longer failing if not provided
<img width="1132" alt="Screenshot 2021-09-01 at 11 12 41" src="https://user-images.githubusercontent.com/73831087/131656735-31f723ca-4a57-4784-ae11-b91a1a8ce5ca.png">
![image](https://user-images.githubusercontent.com/73831087/131656779-a6fcd8e8-08cb-464a-bdea-cd4af05d0176.png)

- I've added information on how to pass the metadata to pcgr-nf on CloudOS  in [docs/usage_cloudos.md](https://github.com/lifebit-ai/pcgr-nf/blob/metadata-dev/docs/usage_cloudos.md) 

Additionally, I think #48 should also be included in this PR so I'm waiting for clarification on that issue. 

Thank you for your review!

Vlad-Dembrovskyi reviewed (approved):
All good now! Without metadata pipeline doesn't fail now
